feat(iota-json-rpc-api, ts-sdk, iota-sdk)!: Introduce `get_dynamic_field_object_v2` RPC APIs

samuel-rufi opened the pull request:
# Description of change

**Feature-branch PR:** Introduces `get_dynamic_field_object_v2` RPC APIs which allow to pass a content option parameter.

Previous PR https://github.com/iotaledger/iota/pull/6316 which added a simple parameter to the existing function signature was closed to avoid any potential breaking changes. For more context on the reasoning see https://github.com/iotaledger/iota/pull/6316#issuecomment-2789572068.

Integration should be done in following places (check means already done):

- [x] iota-sdk
- [x] ts-sdk 

## Links to any relevant issues

Fixes https://github.com/iotaledger/iota/issues/5807
Fixes https://github.com/iotaledger/iota/issues/6116

## Type of change

- Feature (Non-breaking change)

Commit:
feat(iota-json-rpc-api): Introduce `get_dynamic_field_object_v2` RPC APIs (#6331)

# Description of change

Introduces `get_dynamic_field_object_v2` RPC APIs which allow to pass a
content option parameter.

Previous PR https://github.com/iotaledger/iota/pull/6316 which added a
simple parameter to the existing function signature was closed in favor
of this patch to avoid any potential breaking changes.

## Links to any relevant issues

https://github.com/iotaledger/iota/issues/5807

## Type of change

- Feature (Non-breaking change)

## How the change has been tested

```
cargo build
cargo clippy
```
Node RPC tests:
```
cargo nextest run --cargo-profile simulator -p iota-json-rpc-tests
```
Indexer RPC tests:
```
cargo test --profile simulator --features shared_test_runtime
```

## Change checklist

- [X] I have followed the contribution guidelines for this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [X] I have checked that new and existing unit tests pass locally with
my changes

Thoralf-M committed:
feat(iota-sdk): add get_dynamic_field_object_v2() (#6343)

# Description of change

Add get_dynamic_field_object_v2() to the iota-sdk that was introduced to
the RPC API in https://github.com/iotaledger/iota/pull/6331

## Links to any relevant issues

https://github.com/iotaledger/iota/issues/5807

## Type of change

- Feature (Non-breaking change)

## How the change has been tested

```
cargo clippy
```

### Release Notes

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [x] Rust SDK: added `get_dynamic_field_object_v2()`
- [ ] REST API:

begonaalvarezd committed:
Merge branch 'develop' into feat/rpc-get-dynamic-field-object-v2

msarcev committed:
chore(ts-sdk): Support in TS SDK for getDynamicFieldObjectV2 (#6338)

## Links to any relevant issues

Closes https://github.com/iotaledger/iota/issues/6116

DaughterOfMars commented on crates/iota-indexer/src/apis/indexer_api.rs:
```diff
@@ -91,6 +91,65 @@ impl IndexerApi {
+            &name.type_,
+            &name_bcs_value,
+        )
+        .expect("deriving dynamic field id can't fail");
```
If it can't fail then use unwrap and add a comment

kodemartin replied:
The [std](https://doc.rust-lang.org/std/error/index.html#converting-errors-into-panics) docs suggest otherwise

> Of the two, expect is generally preferred since its msg field allows you to convey your intent and assumptions which makes tracking down the source of a panic easier. unwrap on the other hand can still be a good fit in situations where you can trivially show that a piece of code will never panic, such as "127.0.0.1".parse::<std::net::IpAddr>().unwrap() or early prototyping.

DaughterOfMars replied:
Fair enough

DaughterOfMars replied:
Though this is exactly the kind of message that you don't want a user to ever see. 😆 it may be better to say "failed to derive dynamic id" and you could also add a comment saying it shouldn't fail, I guess

kodemartin replied:
That actually raises another issue. @samuel-rufi why don't we just handle the error as an internal server error?

I understand that this `expect` is used as a [precondition](https://doc.rust-lang.org/std/error/index.html#common-message-styles) but a public endpoint does not seem as a place for panicking and developer debugging.

samuel-rufi replied:
Good point, I was only moving the code in a dedicated function and didn't want to touch the logic or change anything in the error handling. I agree, we should return an internal error to the user and log it additionally.

samuel-rufi replied:
Please see https://github.com/iotaledger/iota/pull/6415

DaughterOfMars commented on crates/iota-indexer/src/apis/indexer_api.rs:
```diff
@@ -91,6 +91,65 @@ impl IndexerApi {
+        let options = options.unwrap_or_default();
+
+        match self.inner.get_object_read_in_blocking_task(id).await? {
+            iota_types::object::ObjectRead::NotExists(_)
```
Could import this to shorten it up

samuel-rufi replied:
Please see https://github.com/iotaledger/iota/pull/6415

DaughterOfMars reviewed (commented):
where are the changes between v1 and 2 documented?

Commit:
refactor(iota-indexer): Refactor error handling in Indexer `get_dynamic_field_object_v2` (#6415)

# Description of change

Refactors the error handling in the Indexer
`get_dynamic_field_object_v2` function to return an internal error
instead. Furthermore, it shortens some type usages, relevant imports are
already included.

## Links to any relevant issues

https://github.com/iotaledger/iota/pull/6337#discussion_r2039859668

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Indexer RPC tests:
`cargo test --profile simulator --features shared_test_runtime`

## Change checklist

- [X] I have followed the contribution guidelines for this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [X] I have checked that new and existing unit tests pass locally with
my changes

samuel-rufi commented:
@DaughterOfMars can you be more specific? where should they be documented?

DaughterOfMars commented:
> @DaughterOfMars can you be more specific? where should they be documented?

I'm not sure, but how is a user supposed to know which method to use?

samuel-rufi commented:
@DaughterOfMars in future we probably will depreciate the older one, so this will give additional info. But I guess we need to agree on this and it will require changes from any team that use that API. Currently we only have the OpenRPC which tells the difference between the two methods.
